### PR TITLE
cdc: fix the ordering of batch events

### PR DESCRIPTION
**Pull request by 5kbpers**
Signed-off-by: 5kbpers <tangminghua@pingcap.com>

### What problem does this PR solve?

Problem Summary:
Before sending events to TiCDC, TiKV will cut events into multiple groups that the size of each group must less than 6MB, and the resolved ts event is sent as an individual event.

There is a bug that the ordering between resolved ts events and others would be reversed, that may cause TiCDC panic.

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:
Introduce `EventBatcher` to group events.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

### Release note 
* N/A

### CODE REVIEW

File: components/cdc/src/service.rs [hunk 1]
```diff
@@ -174,38 +234,12 @@ impl ChangeData for Service {
...
-                    match event {
-                        CdcEvent::Event(e) => {
-                            resp_vecs.last_mut().unwrap().mut_events().push(e);
-                            current_events_size += event_size;
-                        }
-                        CdcEvent::ResolvedTs(r) => {
-                            // Set resolved ts as an individual event.
-                            let events = resp_vecs.last_mut().unwrap().take_events();
-                            let mut change_data_event = ChangeDataEvent::default();
-                            change_data_event.set_events(events);
```
Comment (5kbpers) [inline components/cdc/src/service.rs:198]: FYI, here is where the bug happens. 🤥 @overvenus @MyonKeminta

**Review by overvenus — APPROVED** [review]
LGTM

File: components/cdc/src/service.rs [hunk 2]
```diff
@@ -35,18 +35,78 @@ impl ConnID {
...
+                self.last_size += size;
+                self.events.last_mut().unwrap().mut_events().push(e);
+            }
+            CdcEvent::ResolvedTs(r) => {
+                let mut change_data_event = ChangeDataEvent::default();
+                change_data_event.set_resolved_ts(r);
+                self.events.push(change_data_event);
+                // Set last_size to MAX-1 for sending resolved ts as an individual event.
+                // '-1' is to avoid empty event when the next event is still resolved ts.
+                self.last_size = CDC_MAX_RESP_SIZE - 1;
```
Comment (MyonKeminta) [inline components/cdc/src/service.rs:99]: LGTM but somewhat hack..

**Review by MyonKeminta — APPROVED** [review]
LGTM

File: components/cdc/src/service.rs (hunk 2 above)
Comment (5kbpers) [reply components/cdc/src/service.rs:103]: 🤣Yes, but it is a relatively simple way, do you have any suggestion?

Comment (overvenus) [reply components/cdc/src/service.rs:103]: What about adding another flag? 🤔

**Comment by overvenus** [comment]
I'm going to merge this PR, in order to fix daily CI.

I'll send a PR to address https://github.com/tikv/tikv/pull/8629#discussion_r486101231.

**Pull request merged**